extensions/functions/src/main/java/io/crate/window/OffsetValueFunctions.java

```diff
@@ -45,34 +46,173 @@
+            }
+            Object[] rowCells = currentFrame.getRowInPartitionAtIndexOrNull(targetIndex);
+            if (rowCells == null) {
+                throw new IndexOutOfBoundsException();
+            }
+            var row = new RowN(rowCells);
+            for (CollectExpression<Row, ?> expression : expressions) {
+                expression.setNextRow(row);
+            }
+            return args[0].value();
+        }
+    }
+
+    private final Supplier<OffsetDirection> FORWARD = () -> new OffsetDirection() {
+
+        @Override
+        int getTargetIndex(int idxInPartition, int offset) {
+            return idxInPartition + offset;
+        }
+
+        @Override
+        Object getValueAtOffsetIgnoringNulls(int idxInPartition,
+                                             int offset,
+                                             WindowFrameState currentFrame,
+                                             List<? extends CollectExpression<Row, ?>> expressions,
+                                             Input[] args) {
+            if (cachedNonNullIndex == null) {
+                cachedNonNullIndex = findNonNullOffsetFromCurrentIndex(idxInPartition, offset, currentFrame, expressions, args);
+            } else {
+                var curValue = getValueAtTargetIndex(idxInPartition, currentFrame, expressions, args);
+                if (curValue != null) {
+                    moveCacheToNextNonNull(currentFrame, expressions, args);
+                }
+            }
+            return getValueAtTargetIndex(cachedNonNullIndex, currentFrame, expressions, args);
+        }
+    };
+
+    private final Supplier<OffsetDirection> BACKWARD = () -> new OffsetDirection() {
+
+        @Override
+        int getTargetIndex(int idxInPartition, int offset) {
+            return idxInPartition - offset;
+        }
+
+        @Override
+        Object getValueAtOffsetIgnoringNulls(int idxInPartition,
+                                             int offset,
+                                             WindowFrameState currentFrame,
+                                             List<? extends CollectExpression<Row, ?>> expressions,
+                                             Input[] args) {
+            if (cachedNonNullIndex == null) {
+                cachedNonNullIndex = findNonNullOffsetFromCurrentIndex(idxInPartition, offset, currentFrame, expressions, args);
+            } else {
+                var prevValue = getValueAtTargetIndex(idxInPartition - 1, currentFrame, expressions, args);
+                if (prevValue != null) {
+                    moveCacheToNextNonNull(currentFrame, expressions, args);
+                }
+            }
+            return getValueAtTargetIndex(cachedNonNullIndex, currentFrame, expressions, args);
+        }
+    };
+
+    private OffsetDirection resolveOffsetDirection(int offset) {
+        if (LAG_NAME.equals(signature.getName().name())) {
+            if (offset >= 0) {
```

Comment:
Might be possible to parameterize the function with a direction-multiplier to reduce the branching at runtime. Something like this?

```java
return ((defaultDirection * offset) >= 0) ? FORWARD.get() : BACKWARD.get()
```